### ISSUE TITLE
mixing TLS and non-TLS

### ISSUE THREAD

**Issue by haussli** [haussli/draft-dahm-opsawg-tacacs-tls13#7]
The recent changes to the organisation & text have perverted this warning about mixing TLS and non-TLS connections, IMO.

This text
" In order to prevent downgrade attacks, Servers SHOULD keep separate and disjoint lists of clients supporting TLS and Unsecure Connections."

Is this conflating the warning about mixing TLS & non-TLS on clients and servers with the warning about the pitfalls of the servers or clients maintaining a learned list of clients/servers that support (or do not support) TLS[1].  Instead, it should just say that oil and water do not mix.

Perhaps it should be rewritten  roughly as
" In order to prevent downgrade attacks, Servers SHOULD NOT offer both TLS and non-TLS connections on the same server [intentionally not capitalised; meaning "host"] and Clients should not configure both TLS and non-TLS Servers."

Later, the "lists" term is used in this text
"When Migrating from legacy service to TLS, any mixture of Unsecure Connected Servers and TLS-Protected Servers in the same  redundant lists on clients SHOULD be minimised."  and
" After migration, the production deployment SHOULD NOT mix Legacy and TLS-Protected Servers within Server lists configured on clients."

So, perhaps the previous use of the term "lists" is not referring to the warning about "learned lists".  Either way, "lists" seems to be too vague.  I do not have a good suggestion to make it clearer ATM.

[1] removed text was: "Servers and Clients could maintain a cache of Peers that have engaged in TACACS+ TLS connections and demand TLS from that point forward."

**Comment by dcmgashcisco** [comment]
Agreed, the use of the two Lists is ambiguous, I will change the terminology.

One use of List is on the client (i.e. the redundant set of servers it can connect to “within Server lists configured on clients."), the other is on the server (i.e. the list of clients that can connect to it " In order to prevent downgrade attacks, Servers SHOULD keep separate and disjoint lists of clients supporting TLS and Unsecure Connections."). Assuming that we still keep mention of both, I will change so that at least one of the applications does noy use the word “List” to avid the ambiguity.

By the way, I think we may need to tighten the concept of a Server in the document, By Server, do we interpret it as a single port on a TCP address? Or in the classical physical/logical set of ports at an address?

Preventing a server from being able to offer TLS on one port and non TLS on another should not lead to a downgrade attack any more than if the servers were on different addresses as the client will include specification on a port as well as address.

But preventing T+ Server on single port from supporting both would seem like a way to avoid downgrade attack.

The idea of the first “List” here " In order to prevent downgrade attacks, Servers SHOULD keep separate and disjoint lists of clients supporting TLS and Unsecure Connections.", is intended to be a way to prevent this downgrade attack, based on the DB of clients. I’d be happy to remove this approach, it was intended to support Migration but is not critical, and if it avoids an ambiguous “List” all the better.

I think that we discussed a dynamic option: "Servers and Clients could maintain a cache of Peers that have engaged in TACACS+ TLS connections and demand TLS from that point forward." and I thought we had rejected encouraging the creation of such state? I did remove it as I may have erroneously thought it was a discussion point 😊



From: haussli ***@***.***>
Date: Thursday, 13 July 2023 at 21:02
To: haussli/draft-dahm-opsawg-tacacs-tls13 ***@***.***>
Cc: Douglas Gash (dcmgash) ***@***.***>, Assign ***@***.***>
Subject: [haussli/draft-dahm-opsawg-tacacs-tls13] mixing TLS and non-TLS (Issue #7)

The recent changes to the organisation & text have perverted this warning about mixing TLS and non-TLS connections, IMO.

This text
" In order to prevent downgrade attacks, Servers SHOULD keep separate and disjoint lists of clients supporting TLS and Unsecure Connections."

Is this conflating the warning about mixing TLS & non-TLS on clients and servers with the warning about the pitfalls of the servers or clients maintaining a learned list of clients/servers that support (or do not support) TLS[1]. Instead, it should just say that oil and water do not mix.

Perhaps it should be rewritten roughly as
" In order to prevent downgrade attacks, Servers SHOULD NOT offer both TLS and non-TLS connections on the same server [intentionally not capitalised; meaning "host"] and Clients should not configure both TLS and non-TLS Servers."

Later, the "lists" term is used in this text
"When Migrating from legacy service to TLS, any mixture of Unsecure Connected Servers and TLS-Protected Servers in the same redundant lists on clients SHOULD be minimised." and
" After migration, the production deployment SHOULD NOT mix Legacy and TLS-Protected Servers within Server lists configured on clients."

So, perhaps the previous use of the term "lists" is not referring to the warning about "learned lists". Either way, "lists" seems to be too vague. I do not have a good suggestion to make it clearer ATM.

[1] removed text was: "Servers and Clients could maintain a cache of Peers that have engaged in TACACS+ TLS connections and demand TLS from that point forward."

**Comment by dcmgashcisco** [comment]
Agreed, the term "List" is too vague, and used in two different contexts that may cause confusion. I'll look to eliminate this.

The approach I'd have in mind is the following, LMK if this covers your concerns.

1) In the field for non-TLS currently, A TACACS+ Client always defines a TACACS+ Server with both an Address and a Port. AFAIK, no TACACS+ server in the field listens on all ports. By default, for Non TLS, we use port 49, so it is legitimate (though I think not recommended) for a client to specify just an IP address and then the default port 49 would be assumed, but we can say that this is still an implicit server port config.

2) I think for TLS we can mandate the same, but make it more explicit. A TACACS+ Server for TLS is defined by both the address and port. So the server 10.1.1.34:490 is not the same TACACS+ server as 10.1.1.34:49

3) A TACACS+ Server (i.e. the Address:port combination ) MUST not support both TLS and non-TLS. This is the sensible approach. I had in mind that we might want to support migration on one IP/Port combo and hence the need for a list to ensure that a client could not use both during migration, but could use the same server, this is not worthwhile **and I'll withdraw that**. That will eliminate one of the "lists".

4) On the same vein, as a T+ Server (Address:port combination) cannot support TLS and non TLS, then the learned lists would not be needed (which is good as this state is not something that I'd like to encouraged)

5) This still leaves the question: can multiple T+ Servers on the same address (but different ports) have a mixture of TLS and non TLS, with the thought that it can it lead to a downgrade attack. With the above mandate that a TACACS+ server must be specified with address and port, and each address:port combination cannot mix traffic types, then I'm not quite sure how it could be a downgrade attack any more than having a TLS and non TLS on different addresses.

So, I'd propose:

1) I'll tighten up the definition of a T+ Server instnce to indicate its defintion requires both address and port
2) define that a T+ Server can support TLS or non TLS, but not both
3) remove " In order to prevent downgrade attacks, Servers SHOULD keep separate and disjoint lists of clients supporting TLS and Unsecure Connections."
4) tidy up the migration part based on 1-3.

Regarding the downgrade attack, I will re-add that, if we think it is still relevant with the above proposal. Please LMK if you think it is.
